File: src/application/ReviewService.ts
import { Injectable } from "@nestjs/common";
import { ReviewEntity } from "src/infrastructure/entities";
import { EntityManager } from "typeorm";

@Injectable()
export default class ReviewService {
  constructor(private readonly entityManager: EntityManager) {}

  async get(
    product_id: number,
    {
      page = 1,
      perPage = 10,
      sort,
      rating,
    }: {
      page?: number;
      perPage?: number;
      sort?: string;
      rating?: number;
    },
  ) {
    const [field, order] = sort?.split(":") ?? ["created_at", "DESC"];

    const query = this.entityManager
      .getRepository(ReviewEntity)
      .createQueryBuilder("reviews")
      .where("1 = 1")
      .andWhere("reviews.product_id = :product_id", { product_id })
      .andWhere(rating ? "reviews.rating = :rating" : "1=1", { rating })
      .orderBy(`reviews.${field}`, order.toUpperCase() as "ASC" | "DESC")
      .skip((page - 1) * perPage)
      .take(perPage);

    const reviews = await query.getMany();
    const summary = {
      average_rating:
        reviews.map((review) => review.rating).reduce((a, b) => a + b, 0) / reviews.length || 0,
      total_count: reviews.length,
      distribution: {
        1: reviews.filter((review) => review.rating === 1).length,
        2: reviews.filter((review) => review.rating === 2).length,
        3: reviews.filter((review) => review.rating === 3).length,
        4: reviews.filter((review) => review.rating === 4).length,
        5: reviews.filter((review) => review.rating === 5).length,
      },
    };

    return {
      items: reviews,
      summary,
      pagination: {
        total_items: reviews.length,
        total_pages: Math.ceil(reviews.length / perPage),
        current_page: page,
        per_page: perPage,
      },
    };
  }

  async create(product_id: number, review: Partial<ReviewEntity>) {
    const reviewEntity = this.entityManager.create(ReviewEntity, {
      ...review,
      product: { id: product_id },
    });
    return this.entityManager.save(reviewEntity);
  }

  async update(id: number, review: Partial<ReviewEntity>) {
    await this.entityManager.update(ReviewEntity, id, review);

    const updatedReview = await this.entityManager.findOne(ReviewEntity, { where: { id } });
    return updatedReview;
  }

  async delete(id: number) {
    await this.entityManager.delete(ReviewEntity, id);
  }
}

File: src/main/java/com/wanted_preonboarding_challenge_backend/eCommerce/service/ReviewService.java
package com.wanted_preonboarding_challenge_backend.eCommerce.service;

import com.wanted_preonboarding_challenge_backend.eCommerce.dto.common.PaginationDto;
import com.wanted_preonboarding_challenge_backend.eCommerce.domain.Product;
import com.wanted_preonboarding_challenge_backend.eCommerce.domain.Review;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.SummaryDto;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.review.ReviewItemDto;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.review.request.ReviewCreateRequest;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.review.request.ReviewUpdateRequest;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.review.response.ReviewCreateResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.review.response.ReviewGetResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.dto.review.response.ReviewUpdateResponse;
import com.wanted_preonboarding_challenge_backend.eCommerce.mapper.ReviewMapper;
import com.wanted_preonboarding_challenge_backend.eCommerce.respository.ProductRepository;
import com.wanted_preonboarding_challenge_backend.eCommerce.respository.ReviewRepository;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@AllArgsConstructor
public class ReviewService {

    private final ReviewRepository reviewRepository;
    private final ProductRepository productRepository;

    private final ReviewMapper reviewMapper;


    public ReviewGetResponse getReviews(Long id, int page, int perPage, String sort, Integer rating) {
        String[] sortParts = sort.split(":");
        String sortBy = sortParts[0];
        Sort.Direction direction = sortParts.length > 1 && "asc".equalsIgnoreCase(sortParts[1]) ? Sort.Direction.ASC : Sort.Direction.DESC;
        Pageable pageable = PageRequest.of(page - 1, perPage, Sort.by(direction, sortBy));

        Product product = productRepository.findById(id).orElseThrow(() -> new RuntimeException());

        Page<Review> reviewPage = reviewRepository.findByProductAndRating(product, pageable, rating);
        List<Review> allReviews = reviewRepository.findAllByProduct(product);

        List<ReviewItemDto> items = reviewMapper.toReviewItemDtoList(reviewPage.getContent());

        double averageRating = allReviews.stream()
                .mapToInt(Review::getRating)
                .average()
                .orElse(0.0);

        Map<Integer, Long> distribution = allReviews.stream()
                .collect(Collectors.groupingBy(
                        Review::getRating,
                        Collectors.counting()
                ));

        SummaryDto summaryDto = SummaryDto.builder()
                .averageRating(averageRating)
                .totalCount(allReviews.size())
                .distribution(distribution).build();

        PaginationDto paginationDto = PaginationDto.builder()
                .totalItems(reviewPage.getTotalElements())
                .totalPages(reviewPage.getTotalPages())
                .currentPage(reviewPage.getNumber() + 1)
                .perPage(reviewPage.getSize())
                .build();

        ReviewGetResponse response =
                ReviewGetResponse.builder()
                        .items(items)
                        .summary(summaryDto)
                        .pagination(paginationDto)
                        .build();

        return response;
    }

    public void deleteReview(Long reviewId) {

        Review review = reviewRepository.findById(reviewId).orElseThrow(() -> new RuntimeException());

        reviewRepository.delete(review);
    }

    public ReviewCreateResponse writeReview(Long productId, ReviewCreateRequest request) {
        Long userId = 1L;

        Product product = productRepository.findById(productId).orElseThrow(() -> new RuntimeException());

        Review review = reviewMapper.toEntity(request);

        review.writeReview(product);
        reviewRepository.save(review);

        ReviewCreateResponse response = reviewMapper.toReviewCreateResponse(review);
        return response;
    }

    public ReviewUpdateResponse updateReview(Long reviewId, ReviewUpdateRequest request) {

        Review review = reviewRepository.findById(reviewId).orElseThrow(() -> new RuntimeException());

        review.updateReview(review.getRating(), review.getTitle(), review.getContent());

        ReviewUpdateResponse response = reviewMapper.toReviewUpdateResponse(review);

        return response;
    }
}
